### PR TITLE
fix: introduce file cache to simplify fetching files including file mode

### LINKED ISSUE #1027: Do not change modes of existing files
release-please appears to not only create files with, but also change modes of existing files to mode [644](https://github.com/googleapis/release-please/blob/fd724c2a13947fb07de53fd9a9f9798b3b69ee22/src/github.ts#L1401) unconditionally.

This can actually cause breakage -- for example for python projects, `setup.py`, `projectname/__init__.py` and `version.py` found in the tree can be intentionally executable, and making them 644 will break that.

It's fine to create new ones with 644 (although honoring umask would be appropriate as a general practice, haven't checked if it's honored/not), but I suggest making it so that existing files' modes are not changed.

### LINKED ISSUE #1090: Update large files in subdirectories with getFileContentsWithDataAPI (fix created)
#### Environment details

  - OS: Arch Linux (kernel 5.14.3)
  - Node.js version: 14.18.1
  - npm version: 6.14.15
  - `release-please` version: 12.6.0

#### Steps to reproduce

  1. Create a monorepo with a `package.json` and `package-lock.json` in a subdirectory.
  2. Make sure the `package-lock.json` is above the simple API size limit.
  2. Create the release-please manifest and config files for it.
  3. Run `release-please`
  4. It will now throw with `command manifest-pr failed` (see `Output before fix` below)

### The bug
When calling the `getFileContentsOnBranch` function it tries to get the content of specified file with the Github Simple API, but if that file is over the Simple API's size limit, it uses the Data API.

The simple API supports paths to files with several subdirectories, but the Data API only returns one level at the time. So to get a file one or more subdirectories it will need to get the root level, find the correct directory's hash and use this hash to get the next subdirectory or the target file.

Currently it tries to find a path like `subdir1/subdir2/target-file.json` in an array only consisting of the files and directories at root level. It tries to check for a file under a subdirectory with the whole path, but only the first level is returned from the API.

### The fix
I've created a fix with test for it here: [Sherex/release-please](https://github.com/Sherex/release-please).

I do not have an understanding of the whole repo (mostly just the `github.ts` file), so maybe this is redundant and the bug is somewhere else in the repo. But I couldn't find it. :)

I couldn't find a way to get the tree hash of a directory directly through the API, so it steps through the path.

It checks if the path contains a `/` (I assume a name will never contain a slash as it is a path). \[[source](https://github.com/Sherex/release-please/blob/502ec9e9a048de08d060b6fabadeeede6659c533/src/github.ts#L1551)\]  
If it does it splits the path on `/` removes the filename and steps through it one by one. First finding the directory name in the default branch's root tree and takes the SHA1 hash of that directory and requests it's contents.

If it can't find a directory it will throw with `Expected to find a tree (directory) at: ${currentPath}/${directory}`.

When it reaches the end of the array it continues as normal with the last tree, but using just the filename in the find. \[[source](https://github.com/Sherex/release-please/blob/502ec9e9a048de08d060b6fabadeeede6659c533/src/github.ts#L1586)\]  

Let me know if I can create a PR of this (you wanted a issue first according to CONTRIBUTING) or if I should change something. :)

#### Output from run with `--debug`

<details>
  <summary>Output before fix</summary>

```
✔ Found last release sha "**SHA1**" using "bootstrap-sha"
✖ Failed to get .release-please-manifest.json at **SHA1**: 404
✖ Bootstrapping from .release-please-manifest.json at tip of main
✔ Found version 0.1.0 for frontend in .release-please-manifest.json at tip of main
✔ Processing package: Node(package-name)
✔ release as patch: There are 0 BREAKING CHANGES and 4 features
command manifest-pr failed
---------
Error: Could not find requested path: frontend/package-lock.json
    at GitHub.getFileContentsWithDataAPI (/home/sherex/.npm/_npx/410907/lib/node_modules/release-please/build/src/github.js:1324:19)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
    at async /home/sherex/.npm/_npx/410907/lib/node_modules/release-please/build/src/github.js:389:28
    at async GitHub.getFileContentsOnBranch (/home/sherex/.npm/_npx/410907/lib/node_modules/release-please/build/src/github.js:1392:20)
    at async GitHub.getChangeSet (/home/sherex/.npm/_npx/410907/lib/node_modules/release-please/build/src/github.js:1251:41)
    at async Manifest.runReleasers (/home/sherex/.npm/_npx/410907/lib/node_modules/release-please/build/src/manifest.js:297:33)
    at async Manifest.pullRequest (/home/sherex/.npm/_npx/410907/lib/node_modules/release-please/build/src/manifest.js:417:54)
```

</details>

<details>
  <summary>Output after fix</summary>

```
✔ Found last release sha "**SHA1**" using "bootstrap-sha"
✖ Failed to get .release-please-manifest.json at **SHA1**: 404
✖ Bootstrapping from .release-please-manifest.json at tip of main
✔ Found version 0.1.0 for frontend in .release-please-manifest.json at tip of main
✔ Processing package: Node(package-name)
✔ release as patch: There are 0 BREAKING CHANGES and 4 features
❯ Searching "/" for "frontend"
❯ Found it! Getting contents of "/frontend"..
❯ Found 8 files/directories in "/frontend"
✔ updating frontend/package-lock.json from 0.1.0 to 0.1.1
⚠ file frontend/npm-shrinkwrap.json did not exist
⚠ file frontend/samples/package.json did not exist
⚠ frontend/CHANGELOG.md not found
✔ creating frontend/CHANGELOG.md
✔ updating frontend/package.json from 0.1.0 to 0.1.1
✔ Starting GitHub PR workflow...
✔ Successfully found branch HEAD sha "**SHA1 2**".
✔ Skipping branch creation step...
✔ Got the latest commit tree
✔ Successfully created a tree with the desired changes with SHA **SHA1 3**
✔ Successfully created commit. See commit at https://api.github.com/repos/Sherex/package-name-release-please-testing/git/commits/**SHA1 4**
✔ Successfully updated reference release-please/branches/main to **SHA1 4**
✔ Successfully opened pull request available at url: https://api.github.com/repos/Sherex/package-name-release-please-testing/pulls/3.
✔ Successfully opened pull request: 3.
✔ adding label autorelease: pending to https://github.com/Sherex/package-name-release-please-testing/pull/3
```

</details>

### PR DESCRIPTION
**Pull request by chingor13**
This implements a file cache used by the `GitHub` instance. All file fetches will go through this file cache. It uses the git tree API. First, it tries to use the `recursive=true` option to fetch the entire file tree.

If we receive a non-truncated set, we will cache that response and looking all file blob SHAs from that cached tree.

If we receive a truncated set, we will use the tree API without `recursive` and traverse the tree structure until we find the desired node (caching each tree node along the way).

By using the tree API, we can actually fetch the file mode of the fetched file.

Fixes #1090 
Fixes #1027

### CODE REVIEW

**Comment by bcoe** [comment]
Oh cool, so the approach you came up with doesn't use GraphQL, it just pulls the recursive logic for looking up files into a wrapper, and ensures that files are only fetched once.

If I'm understanding correctly you only cache files as they're required, and the idea is that you would be able to get both their contents and mode without two requests?

**Comment by chingor13** [comment]
> If I'm understanding correctly you only cache files as they're required, and the idea is that you would be able to get both their contents and mode without two requests?

We lazy cache git tree data as we need to fetch files. We avoid the simple contents API because we can't get file mode data. We still need separate requests for the file metadata and the actual contents though.

File: test/github.ts [hunk 1]
```diff
@@ -234,57 +220,6 @@ describe('GitHub', () => {
     });
   });
 
-  describe('getFileContentsWithSimpleAPI', () => {
```
[the remark this answers is not in the document]
Comment (chingor13) [reply test/github.ts:237]: Updated to add nock tests for the file cache

**Comment by chingor13** [comment]
> * I'm curious to see how this implementation behaves on a paticularly large repository (could you test against yoshi-go perhaps?).

Running against google-cloud-go seems to still work. That repo is not actually big enough to trigger the recursive file fetching.

File: test/util/file-cache.ts [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+      .get('/repos/testOwner/testRepo/git/trees/dev-branch?recursive=true')
+      .reply(
+        200,
+        require(resolve(
+          fixturesPath,
+          'github-data-api/data-api-trees-successful-response-recursive'
+        ))
+      )
+      .get(
+        '/repos/testOwner/testRepo/git/blobs/3c3629e647f5ddf82548912e337bea9826b434af'
```
Comment (bcoe) [inline test/util/file-cache.ts:389]: awesome, thanks for this.

**Review by bcoe — APPROVED** [review]

**Pull request merged**